File: VB/GridCommander/CommanderTab.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Diagnostics
Imports System.Data
Imports System.IO
Imports System.Text
Imports System.Windows.Forms

Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Views.Grid

Namespace GridCommander
    Partial Public Class CommanderTab
        Inherits DevExpress.XtraEditors.XtraUserControl
        Private _parent As DirectoryInfo
        Private _curentDirectory As DirectoryInfo
        Private Property CurentDirectory() As DirectoryInfo
            Get
                Return _curentDirectory
            End Get
            Set(ByVal value As DirectoryInfo)
                If _curentDirectory Is value Then
                    Return
                End If
                te_CurentText.Text = value.FullName
                _curentDirectory = value
                Dim infos As New List(Of FileSystemInfo)()
                infos.AddRange(_curentDirectory.GetFileSystemInfos())
                If _curentDirectory.Parent IsNot Nothing Then
                    _parent = _curentDirectory.Parent
                    infos.Insert(0, _parent)
                End If
                Grid.DataSource = infos
            End Set
        End Property
        Dim directoryIcon As Icon
        Dim icons As Dictionary(Of String, Icon)
        Public Sub New()
            InitializeComponent()
            directoryIcon = New Icon("../../Resources/Folder-Open.ico")
            icons = New Dictionary(Of String, Icon)
        End Sub

        Private Sub Grid_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Grid.Load
            cmb_Drives.Properties.Items.AddRange(Directory.GetLogicalDrives())
            cmb_Drives.Text = cmb_Drives.Properties.Items(0).ToString()
        End Sub

        Private Sub cmb_Drives_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cmb_Drives.EditValueChanged
            CurentDirectory = New DirectoryInfo(cmb_Drives.Text)
        End Sub

        Private Sub cmb_Drives_EditValueChanging(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ChangingEventArgs) Handles cmb_Drives.EditValueChanging
            Try
                CurentDirectory = New DirectoryInfo(e.NewValue.ToString())
            Catch
                CurentDirectory = New DirectoryInfo(e.OldValue.ToString())
                e.Cancel = True
            End Try
        End Sub

        Private Sub gridView1_DoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles gridView1.DoubleClick
            Dim view = CType(sender, GridView)
            Dim fi As FileSystemInfo = TryCast(view.GetRow(view.FocusedRowHandle), FileSystemInfo)
            If (fi.Attributes And FileAttributes.Directory) <> FileAttributes.Directory Then
                Process.Start(fi.FullName)
            Else
                CurentDirectory = New DirectoryInfo(fi.FullName)
            End If
            Return
        End Sub

        Private Sub CommanderTab_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles Grid.KeyDown, MyBase.KeyDown
            If (e.KeyCode = Keys.Back) AndAlso (CurentDirectory.Parent IsNot Nothing) Then
                CurentDirectory = CurentDirectory.Parent
            End If
        End Sub

        Private Sub gridView1_CustomDrawCell(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs) Handles gridView1.CustomDrawCell
            Dim view = CType(sender, GridView)
            If e.Column Is colName Then
                Dim ic As Icon
                Dim fa As FileAttributes = CType(view.GetRowCellValue(e.RowHandle, colAttributes), FileAttributes)
                If (fa And FileAttributes.Directory) = FileAttributes.Directory Then
                    ic = directoryIcon
                Else
                    Dim path As String = CStr(view.GetRowCellValue(e.RowHandle, colFullName))
                    If Not icons.ContainsKey(path) Then
                        icons.Add(path, Icon.ExtractAssociatedIcon(path))
                    End If
                    ic = icons(path)
                End If
                Dim rec As Rectangle = e.Bounds
                rec.Width = rec.Height
                rec.Inflate(-2, -2)
                rec.Offset(1, 1)
                e.Graphics.DrawIcon(ic, rec)
            End If
        End Sub

        Private Sub gridView1_CustomColumnDisplayText(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDisplayTextEventArgs) Handles gridView1.CustomColumnDisplayText
            Dim view = CType(sender, GridView)
            If e.Column Is colName Then
                Dim s As String = CStr(view.GetListSourceRowCellValue(e.ListSourceRowIndex, colFullName))
                If (_parent IsNot Nothing) AndAlso (s = _parent.FullName) Then
                    e.DisplayText = "[..]"
                Else
                    Dim fa As FileAttributes = CType(view.GetListSourceRowCellValue(e.ListSourceRowIndex, colAttributes), FileAttributes)
                    If (fa And FileAttributes.Directory) = FileAttributes.Directory Then
                        e.DisplayText = "[" & e.Value.ToString() & "]"
                    End If
                End If
            End If
            If e.Column Is colSize Then
                Dim fa As FileAttributes = CType(view.GetListSourceRowCellValue(e.ListSourceRowIndex, colAttributes), FileAttributes)
                If (fa And FileAttributes.Directory) = FileAttributes.Directory Then
                    e.DisplayText = "<DIR>"
                Else
                    Dim path As String = CStr(view.GetListSourceRowCellValue(e.ListSourceRowIndex, colFullName))
                    Dim fi As New FileInfo(path)
                    e.DisplayText = fi.Length.ToString()
                End If
            End If
        End Sub
    End Class
End Namespace

File: CS/GridCommander/CommanderTab.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using System.IO;
using System.Text;
using System.Windows.Forms;

using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Views.Grid;

namespace GridCommander {
    public partial class CommanderTab : DevExpress.XtraEditors.XtraUserControl {
        DirectoryInfo _parent;
        DirectoryInfo _curentDirectory;
        DirectoryInfo CurentDirectory {
            get { return _curentDirectory; }
            set {
                if(_curentDirectory == value)
                    return;
                te_CurentText.Text = value.FullName;
                _curentDirectory = value;
                List<FileSystemInfo> infos = new List<FileSystemInfo>();
                infos.AddRange(_curentDirectory.GetFileSystemInfos());
                if(_curentDirectory.Parent != null) {
                    _parent = _curentDirectory.Parent;
                    infos.Insert(0, _parent);
                }
                Grid.DataSource = infos;
            }
        }
        Icon directoryIcon;
        Dictionary<string, Icon> icons;
        public CommanderTab() {
            InitializeComponent();
            directoryIcon = new Icon("../../Resources/Folder-Open.ico");
            icons = new Dictionary<string, Icon>();
        }

        private void Grid_Load(object sender, EventArgs e) {
            cmb_Drives.Properties.Items.AddRange(Directory.GetLogicalDrives());
            cmb_Drives.Text = cmb_Drives.Properties.Items[0].ToString();
        }

        private void cmb_Drives_EditValueChanged(object sender, EventArgs e) {
            CurentDirectory = new DirectoryInfo(cmb_Drives.Text);
        }

        private void cmb_Drives_EditValueChanging(object sender, DevExpress.XtraEditors.Controls.ChangingEventArgs e) {
            try {
                CurentDirectory = new DirectoryInfo(e.NewValue.ToString());
            }
            catch {
                CurentDirectory = new DirectoryInfo(e.OldValue.ToString());
                e.Cancel = true;
            }
        }

        private void gridView1_DoubleClick(object sender, EventArgs e) {
            GridView view = sender as GridView;
            FileSystemInfo fi = view.GetRow(view.FocusedRowHandle) as FileSystemInfo;
            if((fi.Attributes & FileAttributes.Directory) != FileAttributes.Directory)
                Process.Start(fi.FullName);
            else
                CurentDirectory = new DirectoryInfo(fi.FullName);
        }

        private void CommanderTab_KeyDown(object sender, KeyEventArgs e) {
            if(e.KeyCode == Keys.Back && CurentDirectory.Parent != null)
                CurentDirectory = CurentDirectory.Parent;
        }

        private void gridView1_CustomDrawCell(object sender, DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs e) {
            GridView view = sender as GridView;
            if(e.Column == colName) {
                Icon ic;
                FileAttributes fa = (FileAttributes)view.GetRowCellValue(e.RowHandle, colAttributes);
                if((fa & FileAttributes.Directory) == FileAttributes.Directory) 
                    ic = directoryIcon;
                else { 
                    string path = (string)view.GetRowCellValue(e.RowHandle, colFullName);
                    if(!icons.ContainsKey(path))
                        icons.Add(path, Icon.ExtractAssociatedIcon(path));
                    ic = icons[path];
                }
                Rectangle rec = e.Bounds;
                rec.Width = rec.Height;
                rec.Inflate(-2, -2);
                rec.Offset(1, 1);
                e.Graphics.DrawIcon(ic, rec);
            }
        }

        private void gridView1_CustomColumnDisplayText(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDisplayTextEventArgs e) {
            GridView view = sender as GridView;
            if(e.Column == colName) {
                string s = (string)view.GetListSourceRowCellValue(e.ListSourceRowIndex, colFullName);
                if((_parent != null) && (s == _parent.FullName))
                    e.DisplayText = "[..]";
                else {
                    FileAttributes fa = (FileAttributes)view.GetListSourceRowCellValue(e.ListSourceRowIndex, colAttributes);
                    if((fa & FileAttributes.Directory) == FileAttributes.Directory)
                        e.DisplayText = "[" + e.Value + "]";
                }
            }
            if(e.Column == colSize) {
                FileAttributes fa = (FileAttributes)view.GetListSourceRowCellValue(e.ListSourceRowIndex, colAttributes);
                if((fa & FileAttributes.Directory) == FileAttributes.Directory)
                    e.DisplayText = "<DIR>";
                else {
                    string path = (string)view.GetListSourceRowCellValue(e.ListSourceRowIndex, colFullName);
                    FileInfo fi = new FileInfo(path);
                    e.DisplayText = fi.Length.ToString();
                }
            }
        }
    }
}
